File: Solutions/910-Smallest Range II/910-Smallest Range II.js
//Third fastest
/**
 * @param {number[]} A
 * @param {number} K
 * @return {number}
 */
var smallestRangeII = function(A, K) {
        const N = A.length;
        A.sort((a, b) => a-b);
        let ans = A[N-1] - A[0]; 
    
        for (let i = 0; i < A.length - 1; i++) {
            const h = Math.max(A[N-1] - K, A[i] + K); 
            const l = Math.min(A[0] + K, A[i+1] - K);
            ans = Math.min(ans, h - l);
        }
        return ans;
};

// high is min of (the maxElement - K or current + K)
// low is max of (the first element + K or current - K)
// largest range.

//Second fastest
/**
 * @param {number[]} A
 * @param {number} K
 * @return {number}
 */
var smallestRangeII = function(A, K) {
        const N = A.length;
        A.sort((a, b) => a-b);
        let ans = A[N-1] - A[0]; 
    
        for (let i = 0; i < A.length - 1; i++) {
            const a = A[i], b = A[i+1];
            const high = Math.max(A[N-1] - K, a + K); 
            const low = Math.min(A[0] + K, b - K);
            ans = Math.min(ans, high - low);
        }
        return ans;
};

// high is min of (the maxElement - K or current + K)
// low is max of (the first element + K or current - K)
// largest range.

//Fastest
/**
 * @param {number[]} A
 * @param {number} K
 * @return {number}
 */
var smallestRangeII = function(A, K) {
    if (A.length === 1) return 0
    A.sort((a,b) => a-b);
    let ans = A[A.length - 1] - A[0];
    for (let i = 0; i < A.length - 1; i++) {
        let max = Math.max(A[A.length - 1] - K, A[i] + K);
        let min = Math.min(A[0] + K, A[i + 1] - K)
        ans = Math.min(ans, max -min)
    }
    return ans
};

File: Solutions/910-Smallest Range II/910-Smallest Range II.cs
public class Solution {
    public int SmallestRangeII(int[] A, int K) {
	int N = A.Length;
	Array.Sort(A);
	int ans = A[N-1] - A[0];

	for (int i = 0; i < A.Length - 1; ++`i) {
		int a = A[i], b = A[i+1];
		int high = Math.Max(A[N-1] - K, a + K);
		int low = Math.Min(A[0] + K, b - K);
		ans = Math.Min(ans, high - low);
	}
	return ans;
}
}

//Fastest
public class Solution {
    public int SmallestRangeII(int[] A, int K) {
        Array.Sort(A);
        var result = A.Last() - A.First();
        for (var i = 0; i < A.Length - 1; i++) {
            var localMax = Math.Max(A[i] + K, A.Last() - K);
            var localMin = Math.Min(A[i + 1] - K, A.First() + K);
            result = Math.Min(result, localMax - localMin);
        }
        
        return result;
    }
}